### PR TITLE
Folder: Replace folderId with folderUid

### LINKED ISSUE #57874: Replace the use of folderId with folderUids in library panels, unified alerting and dashboard
- Folder picker in unified alerting
- Returning to view mode after editing a panel
- Folder picker in dashboard general settings
- Saving dashboard
- pre-selecting folder in save dashboard form modal
- uid in routes when opening a dashboard
- saving dashboard json in dashboard settings
- dashboard/folder name validation
- moving dashboard to another folder
- dashboard action buttons
- dashboard creations on an empty folder
- showing library panel
- saving a library panel
- DashList Panel- (won't be doing yet as search service isn't support folderUid when getting the list of dashboards attached to a folder)
- AlertList Panel -  (won't be doing yet as alerting service isn't support folderUid when getting the list of alerts attached to a folder)

### PR DESCRIPTION
**Pull request by lpskdl**
**What is this feature?**

Replace folderId with folderUid in the following areas:
- Folder picker in unified alerting
- Returning to view mode after editing a panel
- Folder picker in dashboard general settings
- Saving dashboard
- pre-selecting folder in save dashboard form modal
- uid in routes when opening a dashboard
- saving dashboard json in dashboard settings
- dashboard/folder name validation
- moving dashboard to another folder
- dashboard action buttons
- dashboard creations on an empty folder
- showing library panel
- saving a library panel

**Why do we need this feature?**
- Prework needed for nested folder feature. Although the backend will still support backward compatibility of using `folderId` it would be easier to start on cleaner state when we start the actual feature as they will be mostly working with `folderUid`s

~~Blocked by  issue #57988. This is needed to avoid supporting `folderId` and  `folderUid` in `ImportDashboardForm` component.~~

**Which issue(s) does this PR fix?**:

Fixes #57874

**Special notes for your reviewer**:
- ~~Import dashboard isn't working at the moment with `folderUid` it will be solve when #57988 is addressed~~
- DashList and AlertList are still using folderId as the api to get those list aren't supporting folderUid at the moment. Separate issue will be created for that and be assigned to the backend.

### CODE REVIEW

**Review by soniaAguilarPeiron — APPROVED** [review]
Looks great from alerting side! 🎉

**Review by axelavargas — APPROVED** [review]
Fantastic work @lpskdl 🥳 , the code looks good to me, I also gave it a try locally, and works as expected!

File: public/app/features/dashboard/state/initDashboard.ts [hunk 1]
```diff
@@ -27,7 +35,8 @@ export interface InitDashboardArgs {
   urlUid?: string;
   urlSlug?: string;
   urlType?: string;
-  urlFolderId?: string;
+  // urlFolderId?: string;
```
[the remark this answers is not in the document]
Comment (lpskdl) [reply public/app/features/dashboard/state/initDashboard.ts]: thanks @axelavargas. I removed it

**Comment by kaydelaney** [comment]
Would it be possible for this to get merged soon? :)

File: public/app/features/manage-dashboards/state/actions.ts [hunk 2]
```diff
@@ -203,13 +203,16 @@ export function moveDashboards(dashboardUids: string[], toFolder: FolderInfo) {
 async function moveDashboard(uid: string, toFolder: FolderInfo) {
   const fullDash: DashboardDTO = await getBackendSrv().get(`/api/dashboards/uid/${uid}`);
 
-  if ((!fullDash.meta.folderId && toFolder.id === 0) || fullDash.meta.folderId === toFolder.id) {
+  if (
+    (fullDash.meta.folderUid === undefined && fullDash.meta.folderId === null && toFolder.uid === '') ||
+    fullDash.meta.folderUid === toFolder.uid
+  ) {
```
Comment (JoaoSilvaGrafana) [inline public/app/features/manage-dashboards/state/actions.ts:209]: Just out of curiosity, which scenarios does `folderUid === undefined && folderId === null`? is it some sort of legacy thing?

**Review by JoaoSilvaGrafana — APPROVED** [review]
Nice one! 👏

File: public/app/features/manage-dashboards/state/actions.ts (hunk 2 above)
Comment (lpskdl) [reply public/app/features/manage-dashboards/state/actions.ts]: Good eyes @JoaoSilvaGrafana it suppose to be `folderUid`. About the scenario, It seems part of a legacy thing, I'm not entirely sure which scenario it catches as it always go to the second condition(`fullDash.meta.folderUid === toFolder.uid`). 

Probably there might be a case where `folderUid` is `undefined` or `null` and if this happen the same as the destination folder is `''` which is general then we assume that they are in the same general folder location hence sending `{ alreadyInFolder: true };`

**Pull request merged**